Add focus outline border to focusable components

jarolrod opened the pull request:
Replaces https://github.com/bitcoin-core/gui-qml/pull/214

Adds the focus outline to the (intended) focusable components:

| a | b | c | d | e | f | g | h | i | j | k |
| - | - | - | - | - | - | - | - | - | - | - |
| <img width="752" alt="Screen Shot 2023-02-13 at 8 48 31 PM" src="https://user-images.githubusercontent.com/23396902/218634844-31045e76-5673-4d74-8606-8c1b27a3cf5e.png"> |  <img width="752" alt="Screen Shot 2023-02-13 at 8 48 40 PM" src="https://user-images.githubusercontent.com/23396902/218634877-f935fbd2-289c-4967-a31e-953daeda08d2.png"> |  <img width="752" alt="Screen Shot 2023-02-13 at 8 48 45 PM" src="https://user-images.githubusercontent.com/23396902/218634925-d42f984e-a58d-43bb-88b3-49de6a273ff0.png"> |  <img width="752" alt="Screen Shot 2023-02-13 at 8 48 53 PM" src="https://user-images.githubusercontent.com/23396902/218634957-7b0d458b-80c2-4e02-938f-20324e2235fa.png"> |  <img width="752" alt="Screen Shot 2023-02-13 at 8 48 55 PM" src="https://user-images.githubusercontent.com/23396902/218634993-ae906fc9-0410-4551-a3a8-890bacea76fc.png"> | <img width="752" alt="Screen Shot 2023-02-13 at 8 48 58 PM" src="https://user-images.githubusercontent.com/23396902/218635019-040cd823-e9f8-4e86-b35b-8f4248b9d6a8.png"> | <img width="752" alt="Screen Shot 2023-02-13 at 8 49 01 PM" src="https://user-images.githubusercontent.com/23396902/218635041-820f0e49-9abc-4b61-8a7b-ec18a927d12f.png"> |  <img width="752" alt="Screen Shot 2023-02-13 at 8 49 19 PM" src="https://user-images.githubusercontent.com/23396902/218635060-3f94b33f-2be2-443b-a681-09ee14e9c467.png"> |  <img width="752" alt="Screen Shot 2023-02-13 at 8 49 21 PM" src="https://user-images.githubusercontent.com/23396902/218635100-1a1aec9b-38c6-4d62-9265-3f9700b07d4e.png"> |  <img width="752" alt="Screen Shot 2023-02-13 at 8 49 26 PM" src="https://user-images.githubusercontent.com/23396902/218635123-ef687083-aee4-4db0-9c1d-7dd8ebfdd75c.png"> | <img width="752" alt="Screen Shot 2023-02-13 at 8 49 31 PM" src="https://user-images.githubusercontent.com/23396902/218635147-f00d6480-337f-4947-88a6-920c9f463d12.png"> |

https://github.com/bitcoin-core/gui-qml/issues/219 is still something we have to fix in order for this to work properly

Additionally, this insets the navbuttons 4px from the borders of the GUI to allow for the focus outline to show properly. I believe that **GBKS** intended the buttons to be inset a bit from the borders

| a | b |
| - | - |
| <img width="752" alt="Screen Shot 2023-02-13 at 8 36 33 PM" src="https://user-images.githubusercontent.com/23396902/218636059-13ecf932-5e41-4bb7-94bb-264523283438.png"> |  <img width="752" alt="Screen Shot 2023-02-13 at 8 38 57 PM" src="https://user-images.githubusercontent.com/23396902/218636186-499c5549-b2cc-4325-aade-022be3f47e85.png"> |


[![Windows](https://img.shields.io/badge/OS-Windows-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/win64/insecure_win_gui.zip?branch=pull/264)
[![Intel macOS](https://img.shields.io/badge/OS-Intel%20macOS-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/macos/insecure_mac_gui.zip?branch=pull/264)
[![Apple Silicon macOS](https://img.shields.io/badge/OS-Apple%20Silicon%20macOS-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/macos_arm64/insecure_mac_arm64_gui.zip?branch=pull/264)
[![ARM64 Android](https://img.shields.io/badge/OS-Android-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/android/insecure_android_apk.zip?branch=pull/264)

GBKS commented:
@jarolrod looks awesome. Yes, the nav buttons are inset from the edges a bit. Looks better, and allows for the focus state to be visible.

jarolrod commented:
Updated from [8e6365b](https://github.com/bitcoin-core/gui-qml/commit/8e6365b296255de755d2a6cbd4f4a6fb00e5120d) to [a14c3a8](https://github.com/bitcoin-core/gui-qml/commit/a14c3a8405db3b971a4660d2a4030908566b6d61), [compare](https://github.com/bitcoin-core/gui-qml/compare/8e6365b296255de755d2a6cbd4f4a6fb00e5120d..a14c3a8405db3b971a4660d2a4030908566b6d61)

changes: rebased over main

hebasto commented:
Once more rebase?

jarolrod commented:
updated from [a14c3a8 ](https://github.com/bitcoin-core/gui-qml/commit/a14c3a8405db3b971a4660d2a4030908566b6d61)to [48635c1](https://github.com/bitcoin-core/gui-qml/commit/48635c1e98a00905907d97424bd7b2af0a725bf5), [compare](https://github.com/bitcoin-core/gui-qml/compare/a14c3a8405db3b971a4660d2a4030908566b6d61..48635c1e98a00905907d97424bd7b2af0a725bf5)

changes: rebased over main

johnny9 commented on src/qml/components/BlockClock.qml:
```diff
@@ -81,6 +81,15 @@ Item {
             nodeModel.pause = root.paused
         }
+        Rectangle {
+            visible: root.visualFocus
```
```suggestion
            visible: root.activeFocus && (root.focusReason == Qt.TabFocusReason || root.focusReason == Qt.BacktabFocusReason)
```

johnny9 replied:
Item doesn't have the visualFocus property.  The suggested change is equivalent.

jarolrod replied:
Thanks! fixed, A follow-up needs to figure out how to enable keyboard click to signal clicked to the mouse area when tab focused

<img width="263" alt="Screen Shot 2023-02-16 at 1 27 07 PM" src="https://user-images.githubusercontent.com/23396902/219455745-f4a052ba-45d6-4c66-b2e9-61c5db0207fb.png">

johnny9 reviewed (changes requested):
BlockClock has a bug that causes the focus outline to always be on.

jarolrod commented:
Updated from [48635c1 ](https://github.com/bitcoin-core/gui-qml/commit/48635c1e98a00905907d97424bd7b2af0a725bf5)to [350bf21](https://github.com/bitcoin-core/gui-qml/commit/350bf21904c02ed3a4aa20b1caf416b5b9aad350), [compare](https://github.com/bitcoin-core/gui-qml/compare/48635c1e98a00905907d97424bd7b2af0a725bf5..350bf21904c02ed3a4aa20b1caf416b5b9aad350)

changes: addressed review [comment](https://github.com/bitcoin-core/gui-qml/pull/264#discussion_r1107941902)

johnny9 reviewed (approved):
ACK 350bf21904c02ed3a4aa20b1caf416b5b9aad350

jarolrod commented:
Updated from[ 350bf21 ](https://github.com/bitcoin-core/gui-qml/commit/350bf21904c02ed3a4aa20b1caf416b5b9aad350)to [bae7abe](https://github.com/bitcoin-core/gui-qml/commit/bae7abe37b553c11976927933344e001f25fc046), [compare](https://github.com/bitcoin-core/gui-qml/compare/350bf21904c02ed3a4aa20b1caf416b5b9aad350..bae7abe37b553c11976927933344e001f25fc046)

Changes: changed approach so that
- The Outline Rectangle is a reusable component called FocusBorder
- This FocusBorder encapsulates property definitions as well as behavior animations to reduce code
- Handle background states on tab focus for relevant controls

jarolrod commented:
Updated from [bae7abe ](https://github.com/bitcoin-core/gui-qml/commit/bae7abe37b553c11976927933344e001f25fc046)to [6d43e20](https://github.com/bitcoin-core/gui-qml/commit/6d43e205da965c224a83c67e23bc6c1a1ea659e5), [compare](https://github.com/bitcoin-core/gui-qml/compare/bae7abe37b553c11976927933344e001f25fc046..6d43e205da965c224a83c67e23bc6c1a1ea659e5)

changes: fixed no newline at elf

johncarl-te commented on src/qml/controls/NavButton.qml:
```diff
@@ -103,4 +107,11 @@ AbstractButton {
         }
     }
+    onVisualFocusChanged: {
+        if (visualFocus && (focusReason == Qt.TabFocusReason || focusReason == Qt.BacktabFocusReason)) {
```
Any reason for the focusReason checks? visualFocus implies TabFocusReason.

johnny9 commented on src/qml/controls/TextButton.qml:
```diff
@@ -78,4 +82,12 @@ Button {
     }
+
+    onVisualFocusChanged: {
+        if (visualFocus && (focusReason == Qt.TabFocusReason || focusReason == Qt.BacktabFocusReason)) {
```
visualFocus implies focusReason == Qt.TabFocus. Any reason for the redundant check?

jarolrod replied:
thanks, seems like it can actually be simplified into something like

``` diff
diff --git a/src/qml/controls/ContinueButton.qml b/src/qml/controls/ContinueButton.qml
index a49864075..6c51b40bf 100644
--- a/src/qml/controls/ContinueButton.qml
+++ b/src/qml/controls/ContinueButton.qml
@@ -27,15 +27,15 @@ Button {
 
         states: [
             State {
-                name: "DEFAULT"
+                name: "DEFAULT"; when: !visualFocus
                 PropertyChanges { target: bg; color: Theme.color.orange }
             },
             State {
-                name: "HOVER"
+                name: "HOVER"; when: visualFocus
                 PropertyChanges { target: bg; color: Theme.color.orangeLight1 }
             },
             State {
-                name: "PRESSED"
+                name: "PRESSED"; when: clicked
                 PropertyChanges { target: bg; color: Theme.color.orangeLight2 }
             }
         ]
@@ -65,11 +65,4 @@ Button {
             root.clicked()
         }
     }
-    onVisualFocusChanged: {
-        if (visualFocus && (focusReason == Qt.TabFocusReason || focusReason == Qt.BacktabFocusReason)) {
-            root.background.state = "HOVER"
-        } else {
-            root.background.state = "DEFAULT"
-        }
-    }
 }
```

will update

jarolrod commented:
Updated from[ 6d43e20 ](https://github.com/bitcoin-core/gui-qml/commit/6d43e205da965c224a83c67e23bc6c1a1ea659e5)to[ ab0ea57](https://github.com/bitcoin-core/gui-qml/commit/ab0ea572ac70ffe44e2ca4c90ec9dca826faf83f), [compare](https://github.com/bitcoin-core/gui-qml/compare/6d43e205da965c224a83c67e23bc6c1a1ea659e5..ab0ea572ac70ffe44e2ca4c90ec9dca826faf83f)

Changes:
- dropped changes on checking the reasons behind the visual focus change
- No longer change backgrounds of certain components when given focus per feedback on slack
- squashed commits related to adding the usage of FocusBorder into one commit as the changes are now simple enough.

johnny9 reviewed (approved):
ACK ab0ea572ac70ffe44e2ca4c90ec9dca826faf83f

promag commented on src/qml/controls/FocusBorder.qml:
```diff
@@ -0,0 +1,35 @@
+        ColorAnimation { duration: 150 }
+    }
+
+    Behavior on visible {
+        NumberAnimation { duration: 150 }
+    }
```
What is the idea? should it be opacity animation?

GBKS replied:
Yes, a simple opacity animation is fine.
